src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -33,20 +34,37 @@ public async Task<SolutionInfo> CreateSolutionInfoAsync(Checksum solutionChecksu
         var solutionAttributes = await GetAssetAsync<SolutionInfo.SolutionAttributes>(AssetPathKind.SolutionAttributes, solutionChecksums.Attributes, cancellationToken).ConfigureAwait(false);
         await GetAssetAsync<SourceGeneratorExecutionVersionMap>(AssetPathKind.SolutionSourceGeneratorExecutionVersionMap, solutionCompilationChecksums.SourceGeneratorExecutionVersionMap, cancellationToken).ConfigureAwait(false);
 
-        using var _ = ArrayBuilder<ProjectInfo>.GetInstance(solutionChecksums.Projects.Length, out var projects);
-        foreach (var (projectChecksum, projectId) in solutionChecksums.Projects)
-            projects.Add(await CreateProjectInfoAsync(projectId, projectChecksum, cancellationToken).ConfigureAwait(false));
+        // Fetch all the project state checksums up front.  That allows gettign all the data in a single call, and
+        // enables parallel fetching of the projects below.
+        using var _1 = ArrayBuilder<ProjectStateChecksums>.GetInstance(solutionChecksums.Projects.Length, out var allProjectStateChecksums);
```

Comment:
i could likely make this into the array of projectinfo tasks and populate it as we're reading back the ProjectStateChecksums.